Lwm2m bootstrap fixes

SeppoTakalo opened the pull request:
This PR fixes major bootstrap related issue and few minor issues.

## Bootstrap issue

After introducing a tickless state-machine, our client now enters BOOTSTRA_TRANS_DONE immeadialy when Bootstrap-finnish packet is received. The state transition causes existing socket to be closed, and enters a next state which is `DO_REGISTRATION`. The problem is that ACK packet to Finnish message was still on the send-queue.
This is a problem on some bootstrap server, as the Finnish is confirmable message, and it may consider bootstrap to be failed as we just closed the connection without ACK.
![Screenshot from 2023-08-10 12-53-12](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/2f592ed8-cebe-49fa-9e85-16a18263fede)

Fix the issue by introducing `sm_set_state_delayed()` function that causes the next state transition to be delayed. The same delay was previously a side, effect of slow socket-loop. (There is a random possibility, that it did not work previously, but Wireshark shows that it does).

This fix alone was not enough, as some callback modified the next event timestamp directly.

## Do not directly modify state variables

On many places in RD client. `ctx.client_state` was directly written, followed by call `next_event_at(0)`.
I removed all these, and refactored it to use `set_sm_state()`. Notable change was the `lwm2m_rd_client_connection_resume()` as it modified the timestamp unconditionally, and without this fix, the previous one does not work.

## Wake up engine on pause/resume

Engine pause and Engine resume are public API, so instead of waiting for event-loop to wake up, they should cause it to immediately wake up, so there would be no delay.

This is minor fix, and nobody was complaining it yet.

SeppoTakalo committed:
net: lwm2m: Wake up engine on pause/resume

Engine wake-up call was missing from pause/resume APIs
which caused delay.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lwm2m: Do not immediately close connection at Bootstrap

When BOOTSTRAP FINNISH message was received, it caused
engine to immediately switch to BOOTSTRAP_TRANS_DONE state
which then closed the connection.
Ack packet was still on the send-queue so it never got send before close().

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lwm2m: All state changes should go through set_sm_state()

Some state changes were do by directly writing into
client.engine_state variable, followed by call next_event_at(0);
This causes hard-to-find side effects.

Refactor all state transitions to use set_sm_state() to have better
control for it.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

jheiskan81 reviewed (approved):
Thanks for this fix.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -837,7 +844,7 @@ static int sm_do_bootstrap_reg(void)
 {
 	LOG_INF("Bootstrap data transfer done!");
-	set_sm_state(ENGINE_BOOTSTRAP_TRANS_DONE);
+	set_sm_state_delayed(ENGINE_BOOTSTRAP_TRANS_DONE, 1000);
```
Shall we add a short comment here why this particular state transition is delayed? It might not be obvious and tracking the change commit to get an explanation might be cumbersome.

SeppoTakalo replied:
Sure.
Added a comment.

rlubos reviewed (commented):
Looks good, just one minor suggestion. It also seems that tests need some attention, please take a look.

SeppoTakalo commented:
Fixes the issues found on Twister tests.
Changes in both `_resume()` function caused too many event callbacks, so those have to be the exception that we don't call `sm_set_state()`

SeppoTakalo commented:
Here are better examples of the problem using `native_posix` and locally running Leshan bootstrap server:

## Without the fix
### Leshan output
![Screenshot from 2023-08-11 12-06-36](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/ec6746b4-d5f4-440a-bcb7-1dbda1c80699)
### Wireshark capture
No ACK packet to CoAP POST "/bs" which is the BootstrapFinnish command
![Screenshot from 2023-08-11 12-05-48](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/23b7a878-a622-4765-a983-0a41928379e6)

## With the fix
### Leshan output
No warnings
![Screenshot from 2023-08-11 12-15-22](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/39804382-b65f-475f-b76e-c9a6c83630e4)
### Wireshark capture
Ack normally to last `POST /bs` followed by closing of DTLS with Alert.
![Screenshot from 2023-08-11 12-14-58](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/edd422b4-1415-4a79-95f4-c4271c7ec0b8)

SeppoTakalo committed:
net: lwm2m: Add delay also to exec callbacks

Bootstrap executable cause socket to be closed, so
it requires some delay for Ack to be send out as well.
Similarly, add small delay for Update trigger, so the
possible Ack goes out before the actual update message.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
test: lwm2m: Allow longer waits and stop properly

RD-client tests should wait longer for expected events.
Also the stubbed service worker should be stopped properly
after tests.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
Commits rebased on top of main.
Also, I added two more fixes that I did not realize initially.

When Bootstrap is triggered with executable callback, same issue happens.
Fixed that as well.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -464,7 +466,7 @@ int engine_trigger_bootstrap(void)
 	client.use_bootstrap = true;
 	client.trigger_update = false;
-	set_sm_state(ENGINE_INIT);
+	set_sm_state_delayed(ENGINE_INIT, 1000);
```
Magic numbers should be avoided, we could have here for example `MSEC_PER_SEC` instead

SeppoTakalo replied:
Thanks. Fixed.
I converted all `1000` values on calculations to `MSEC_PER_SEC`.
Then I defined two more macros, `DELAY_BEFORE_CLOSING` and `DELAY_FOR_ACK`

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -837,7 +844,10 @@ static int sm_do_bootstrap_reg(void)
+	/* Delay the state transition, so engine have some time to send ACK
+	 * before we close the socket
+	 */
+	set_sm_state_delayed(ENGINE_BOOTSTRAP_TRANS_DONE, 1000);
```
Magic constant here too
